pkg/admission/prep_downscale_test.go

```diff
@@ -0,0 +1,171 @@
+package admission
+
+import (
+	"bytes"
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"testing"
+	"text/template"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestUpscale(t *testing.T) {
+	testPrepDownscaleWebhook(t, 2, 5, http.StatusOK, true, false)
+}
+
+func TestNoReplicasChange(t *testing.T) {
+	testPrepDownscaleWebhook(t, 5, 5, http.StatusOK, true, false)
+}
+
+func TestDownscaleValidDownscale(t *testing.T) {
+	testPrepDownscaleWebhook(t, 5, 3, http.StatusOK, true, true)
+}
+
+func TestDownscaleInvalidDownscale(t *testing.T) {
+	testPrepDownscaleWebhook(t, 5, 3, http.StatusInternalServerError, false, false)
+}
+
+func newDebugLogger() log.Logger {
+	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+	var options []level.Option
+	options = append(options, level.AllowDebug())
+	logger = level.NewFilter(logger, options...)
+	logger = log.With(logger, "ts", log.DefaultTimestampUTC)
+	return logger
+}
+
+type templateParams struct {
+	Replicas         int
+	DownScalePathKey string
+	DownScalePath    string
+	DownScalePortKey string
+	DownScalePort    string
+}
+
+func testPrepDownscaleWebhook(t *testing.T, oldReplicas, newReplicas, httpStatusCode int, allowed bool, podsPrepared bool) {
```

Comment:
Can we write an integration test please?

Reply:
As this PR is getting bigger and bigger I propose to do this in a separate PR.